File: sdks/typescript/src/transport.ts
/** Current bounded Fetch response carriage and envelope decoding. */

import { type } from "arktype";
import { RrdApiError, RrdClientError } from "./error.js";
import type { OperationId, RequestContext, SuccessPayload } from "./operation.js";

const apiErrorBody = type({
  code: "string",
  message: "string",
  retryable: "boolean",
  "details?": "object",
  "+": "reject",
});
const responseOutcome = type.or(
  type({ status: "'ok'", payload: "unknown", "+": "reject" }),
  type({ status: "'error'", error: apiErrorBody, "+": "reject" }),
);
const responseEnvelope = type({
  protocol: "'rrd'",
  protocol_version: "1",
  request_id: "string",
  operation_id: "string",
  outcome: responseOutcome,
  "+": "reject",
});

export async function decodeResponse<K extends OperationId>(
  response: Response,
  maximumBytes: number,
  expected: RequestContext | undefined,
): Promise<SuccessPayload<K>> {
  const encoded = await readBounded(response, maximumBytes);
  let decoded: unknown;
  try {
    decoded = JSON.parse(encoded);
  } catch (error) {
    throw new RrdClientError(
      error instanceof Error
        ? `RRD response JSON is invalid: ${error.message}`
        : "RRD response JSON is invalid",
    );
  }
  const validated = responseEnvelope(decoded);
  if (validated instanceof type.errors) {
    throw new RrdClientError(`RRD response envelope is invalid: ${validated.summary}`);
  }
  if (
    expected &&
    (validated.request_id !== expected.request_id ||
      validated.operation_id !== expected.operation_id)
  ) {
    throw new RrdClientError("RRD response request/operation identity differs");
  }
  if (response.ok !== (validated.outcome.status === "ok")) {
    throw new RrdClientError("RRD HTTP status and typed outcome disagree");
  }
  if (validated.outcome.status === "error") {
    throw new RrdApiError(response.status, validated.outcome.error);
  }
  return validated.outcome.payload as SuccessPayload<K>;
}

async function readBounded(response: Response, maximum: number): Promise<string> {
  const declared = Number(response.headers.get("content-length"));
  if (Number.isFinite(declared) && declared > maximum) {
    throw new RrdClientError("RRD response exceeded the configured byte limit");
  }
  if (!response.body) return "";
  const reader = response.body.getReader();
  const chunks: Uint8Array[] = [];
  let length = 0;
  for (;;) {
    const { done, value } = await reader.read();
    if (done) break;
    length += value.byteLength;
    if (length > maximum) {
      await reader.cancel();
      throw new RrdClientError("RRD response exceeded the configured byte limit");
    }
    chunks.push(value);
  }
  const combined = new Uint8Array(length);
  let offset = 0;
  for (const chunk of chunks) {
    combined.set(chunk, offset);
    offset += chunk.byteLength;
  }
  return new TextDecoder("utf-8", { fatal: true }).decode(combined);
}

File: sdks/go/transport.go
package rrd

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
)

type responseEnvelope struct {
	Protocol        string          `json:"protocol"`
	ProtocolVersion int             `json:"protocol_version"`
	RequestID       string          `json:"request_id"`
	OperationID     string          `json:"operation_id"`
	Outcome         responseOutcome `json:"outcome"`
}

type responseOutcome struct {
	Status  string          `json:"status"`
	Payload json.RawMessage `json:"payload,omitempty"`
	Error   *APIError       `json:"error,omitempty"`
}

func decodeResponse(
	status int,
	encoded []byte,
	requestContext *requestEnvelopeContext,
) (map[string]any, error) {
	var envelope responseEnvelope
	if err := decodeStrict(encoded, &envelope); err != nil {
		return nil, fmt.Errorf("RRD response envelope is invalid: %w", err)
	}
	if envelope.Protocol != "rrd" || envelope.ProtocolVersion != 1 {
		return nil, errors.New("RRD response protocol differs")
	}
	if requestContext != nil && (envelope.RequestID != requestContext.RequestID ||
		envelope.OperationID != requestContext.OperationID) {
		return nil, errors.New("RRD response request/operation identity differs")
	}
	success := status >= 200 && status < 300
	if success != (envelope.Outcome.Status == "ok") {
		return nil, errors.New("RRD HTTP status and typed outcome disagree")
	}
	if envelope.Outcome.Status == "error" {
		if envelope.Outcome.Error == nil || envelope.Outcome.Payload != nil {
			return nil, errors.New("RRD error outcome is missing its error")
		}
		if envelope.Outcome.Error.Code == "" || envelope.Outcome.Error.Message == "" {
			return nil, errors.New("RRD error outcome is incomplete")
		}
		if envelope.Outcome.Error.Details == nil {
			envelope.Outcome.Error.Details = map[string]string{}
		}
		envelope.Outcome.Error.Status = status
		return nil, envelope.Outcome.Error
	}
	if envelope.Outcome.Status != "ok" {
		return nil, errors.New("RRD response outcome status is invalid")
	}
	if envelope.Outcome.Error != nil || envelope.Outcome.Payload == nil {
		return nil, errors.New("RRD success outcome is invalid")
	}
	var payload map[string]any
	if err := decodeStrict(envelope.Outcome.Payload, &payload); err != nil || payload == nil {
		return nil, errors.New("RRD success payload must be an object")
	}
	return payload, nil
}

func decodeStrict(encoded []byte, output any) error {
	decoder := json.NewDecoder(bytes.NewReader(encoded))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(output); err != nil {
		return err
	}
	if err := decoder.Decode(&struct{}{}); err != io.EOF {
		return errors.New("JSON contains trailing data")
	}
	return nil
}

func readBounded(response *http.Response, maximum int64) ([]byte, error) {
	defer response.Body.Close()
	if response.ContentLength > maximum {
		return nil, errors.New("RRD response exceeded the configured byte limit")
	}
	encoded, err := io.ReadAll(io.LimitReader(response.Body, maximum+1))
	if err != nil {
		return nil, fmt.Errorf("read RRD response: %w", err)
	}
	if int64(len(encoded)) > maximum {
		return nil, errors.New("RRD response exceeded the configured byte limit")
	}
	return encoded, nil
}
